### DISCUSSION TITLE
Is there a way for a script to distinguish dosemu revisions with 4 system drives as opposed to those with 2 system drives?

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1456] [General]
Due to testing [the interrupt 06h fix](https://github.com/dosemu2/dosemu2/commit/65853ec95e78d174964238826c859838e1bad801) I rebuilt dosemu2 from git today. I noticed that there are now only 2 system drives, C: and D:. That means in my ldosboot tests when using a hdimage I need to install the loader to drive E: instead of G:, and to boot it either chainload hdc instead of hde from the debugger (lDebug), or directly boot with `-C2` instead of `-C4`.

I implemented [an alternative machine type](https://hg.ulukai.org/ecm/ldosboot/rev/0a0f49e6ac39) for the 2-drives case for now. However, ideally I'd like the script to test which variant to choose automatically.

**Comment by stsp** [comment]
What are the 2 system drives?
D always contained dosemu tools.

**Reply by ecm-pushbx** [reply]
C: and D:. The prior dosemu2 builds had C:, D:, E:, F: as system drives. Here's the `lredir` output of an older revision:

```
Welcome to dosemu2!
    Build 2.0pre8-20190827-1234-g0984731c9
C:\>lredir
Current Drive Redirections:
C: = \\LINUX\FS\[...]\.dosemu\drive_c\ attrib = READ/WRITE
D: = \\LINUX\FS\[...]\local\share\dosemu\dosemu2-cmds-0.2\ attrib = READ/WRITE
E: = \\LINUX\FS\[...]\local\share\dosemu\fdboot\ attrib = READ/WRITE
F: = \\LINUX\FS\[...]\local\share\dosemu\freedos\ attrib = READ/WRITE
C:\>
```

**Reply by stsp** [reply]
I suppose if you boot from dosemu-freedos-bin,
you will still have all 4 drives.
And if you boot fdpp, you will have
3: one for comcom32.
So while it may depend on version,
it is much more dependent on a config.

**Reply by ecm-pushbx** [reply]
Okay. So how should I detect which config is in use?

**Reply by stsp** [reply]
Not sure.
What do you need?
Fixed drive letter?
Put your stuff on J: or alike.

**Reply by ecm-pushbx** [reply]
I'm using -K to access an executable in the unix cwd, and `-I "disk { hdimage ... }"` as the target of the instsect run. Here: https://hg.ulukai.org/ecm/ldosboot/file/0a0f49e6ac39/test/test.sh#l333

So I need to know which drive letter is assigned to the hdimage, and for the -E command to find its executable in the -K directory's drive.

Further, to boot, I need to know which BIOS unit to load from, which also boils down to knowing how many units there are before the `-I "disk { hdimage ... }"` unit. https://hg.ulukai.org/ecm/ldosboot/file/0a0f49e6ac39/test/test.sh#l400

To boot directly I need `-C4` if there are four units before mine, or `-C2` if there are only two units: https://hg.ulukai.org/ecm/ldosboot/file/0a0f49e6ac39/test/test.sh#l411

**Reply by andrewbird** [reply]
Can't you define the hdimage in dosemu.conf to place your drive before the changing ones? i.e.
~~~
$_hdimage = "+0 special.img +1"
~~~
So your drive always ends up as `D:` no matter how many drives are created by the system.

**Reply by andrewbird** [reply]
I just read your reply, not sure if my hint helps or not now :-)

**Reply by ecm-pushbx** [reply]
I still need to boot FreeDOS at least for running instsect though (in the case with one hdimage drive and one -K drive), and your suggestion seems like it would mess up the FreeDOS boot.

Besides, how do I specify a custom dosemu.conf to dosemu?

**Reply by andrewbird** [reply]
Freedos autoexec and config are hardcoded to `D:`, so yes it would mess that up.

> Besides, how do I specify a custom dosemu.conf to dosemu?

~~~
-n -f mydosemu.conf
~~~

**Reply by stsp** [reply]
> I still need to boot FreeDOS at least for running instsect though

Freedos is currently installed
with the installer scripts - to drive C.
The problem you mention, applies
to deprecated dosemu-freedos-bin.
Fixed drives letter is the only way
forward for you, and the suggestion
to put hdimage on a fixed location
should work.

**Reply by stsp** [reply]
> Put your stuff on J: or alike.

Though I guess your problem is
that there is currently no way to
do that reliably. :)
So will adding some syntax to
$_hdimage to allow a fixed letter,
help?

**Reply by stsp** [reply]
No, it won't help, too, as that would
be fdpp-specific.

**Reply by stsp** [reply]
So how about traversing the
CDS array and find the hdimage
by not having a network flag set?
I mean, when you solve such
task in DOS, you definitely do not
expect any fixed drive or bios unit,
so I suppose its just not a dosemu
problem. Suppose some user always
re-plugs various hard drives. :)
Your program should still work.

**Reply by stsp** [reply]
You can identify the -d/-K
drives in a same way: it
will have the network flag set
and physical flag clear, whereas
normal dosemu drives have
both set.

**Reply by stsp** [reply]
Hmm, actually I recalled an
interesting trick. You can do:
`$_hdimage = "+0 +2 special.img"`
That will put your hdimage to
the fixed location, and will not
affect freedos installed with the
script (but will break dosemu-freedos-bin).

**Reply by ecm-pushbx** [reply]
What does +0 +2 mean? Why would it not work for dosemu-freedos-bin?

**Reply by stsp** [reply]
> What does +0 +2 mean?

It means drives C and D (and not more).

> Why would it not work for dosemu-freedos-bin?

Because it also needs E and F
(that are provided by +1).

**Reply by stsp** [reply]
Thought I am not sure what problem
do we solve. `$_hdimage = "+0 +2 special.img"`
will not work with fdpp, so likely not
what you want?

**Reply by stsp** [reply]
If you want a single config for
any DOS, you need to traverse
the CDS array. If you can have
a custom config per every DOS,
then you always know what drive
will you get with this_config+this_dos.

**Reply by stsp** [reply]
Or if you want a custom config
shared by freedos and fdpp, then
we can have something like:
`$_hdimage = "+0 +2 special.img +3"`
which will work for fdpp and script-installed
freedos.

**Comment by ecm-pushbx** [comment]
I [added some magic detection](https://hg.ulukai.org/ecm/ldosboot/rev/91626ef955f8) now which works with both of my dosemu revisions and either 4 or 2 system drives.
